specreduce/wavelength_calibration.py

```diff
@@ -0,0 +1,303 @@
+        # either the pixel value or a CalibrationLine object with pixel changed to the refined value
+        if self.refinement_method is None:
+            return self.pixel
+        input_spectrum = self.input_spectrum if input_spectrum is None else input_spectrum
+        refined_pixel = self._do_refinement(input_spectrum)
+        if return_object:
+            return CalibrationLine(input_spectrum, self.wavelength, refined_pixel,
+                                   refinement_method=self.refinement_method,
+                                   refinement_kwargs=self.refinement_kwargs)
+
+        return refined_pixel
+
+    @cached_property
+    def refined_pixel(self):
+        # returns the refined pixel for self.input_spectrum
+        return self.refine()
+
+    @cached_property
+    def with_refined_pixel(self):
+        # returns a copy of this object, but with the pixel updated to the refined value
+        return self.refine(return_object=True)
+
+    @property
+    def input_spectrum(self):
+        return self._input_spectrum
+
+    @input_spectrum.setter
+    def input_spectrum(self, new_spectrum):
+        # We want to clear the refined locations if a new calibration spectrum is provided
+        self._clear_cache()
+        self._input_spectrum = new_spectrum
+
+    def __str__(self):
+        return f"CalibrationLine: ({self.wavelength}, {self.pixel})"
+
+    def __repr__(self):
+        return f"CalibrationLine({self.wavelength}, {self.pixel})"
+
+
+class WavelengthCalibration1D():
+
+    def __init__(self, input_spectrum, lines, model=Linear1D(), spectral_unit=u.Angstrom,
+                 fitter=None, default_refinement_method=None, default_refinement_kwargs={}):
+        """
+        input_spectrum: `~specutils.Spectrum1D`
+            A one-dimensional Spectrum1D calibration spectrum from an arc lamp or similar.
+        lines: str, list
+            List of lines to anchor the wavelength solution fit. List items are coerced to
+            CalibrationLine objects if passed as tuples of (wavelength, pixel) with
+            default_refinement_method and default_refinement_kwargs as defaults.
+        model: `~astropy.modeling.Model`
+            The model to fit for the wavelength solution. Defaults to a linear model.
+        default_refinement_method: str, optional
+            words
+        default_refinement_kwargs: dict, optional
+            words
+        """
+        self._input_spectrum = input_spectrum
+        self._model = model
+        self.spectral_unit = spectral_unit
+        self.default_refinement_method = default_refinement_method
+        self.default_refinement_kwargs = default_refinement_kwargs
+        self._cached_properties = ['wcs',]
+
+        if fitter is None:
+            if self.model.linear:
+                self.fitter = LinearLSQFitter(calc_uncertainties=True)
+            else:
+                self.fitter = LMLSQFitter(calc_uncertainties=True)
```

Comment:
what if instead `self._fitter` was stored as `None` and then a `fitter` property defaults based on the _current_ value of `self.model` (so that if `model` is changed after initialization, the default can change to match)?

Reply:
I like this idea, I'll implement momentarily.